### ISSUE TITLE
Transfer Learning YOLO

### ISSUE THREAD

**Issue by monteksingh** [eriklindernoren/PyTorch-YOLOv3#133]
I am trying to set the 'requires_grad' property of the specific layers to FALSE to perform Transfer Learning.

But when I introduce the code below in Training.py-

```
for key, values in model.state_dict().items():
        print(model.state_dict()[key].requires_grad,"Key-",key)
```

I get the below Output-
```
False Key- module_list.0.conv_0.weight
False Key- module_list.0.batch_norm_0.weight
```

i.e 'False' for all the layers even after->  model.train()

I tried to change using ->
Source - https://github.com/ultralytics/yolov3/wiki/Example:-Transfer-Learning

    for i, (name, p) in enumerate(model.named_parameters()):
        if (p.shape[0] == 69):
            p.requires_grad = True
            print(p.requires_grad,"pp",name)
        else:
            p.requires_grad = False

69 because I have (4+1+18classes)  * 3anchors
Changes are shown in-> ` print(p.requires_grad,"pp",name)
`But even after this when I print the 'requires_grad' for individual Layers the output remains FALSE.

```
for key, values in model.state_dict().items():
        print(model.state_dict()[key].requires_grad,"Key-",key)

```
Please explain where am I going wrong.......

**Comment by fourth-archive** [comment]
Just use the repo you cited with `python3 train.py --transfer`. How much easier can it get?
https://github.com/ultralytics/yolov3/wiki/Example:-Transfer-Learning

> This guide explains how to train your data with YOLOv3 **using Transfer Learning**. Transfer learning can be a useful way to quickly retrain YOLOv3 on new data without needing to retrain the entire network. We accomplish this by starting from the official YOLOv3 weights, and setting each layer's `.requires_grad` field to false that we do not want to calculate gradients for and optimize.
> 
> ## Before You Start
> 1. Update software (Python >= 3.7, Pytorch >= 1.0, etc.)
> 2. Clone latest repo: `git clone https://github.com/ultralytics/yolov3`
> 3. Download COCO dataset: `bash yolov3/data/get_coco_dataset.sh`
> 
> ## Transfer Learning
> 1. **Download the official YOLOv3 weights** `yolov3.pt` (converted to PyTorch) from our [Google Drive folder](https://drive.google.com/drive/folders/1uxgUBemJVw9wZsdpboYbzUN4bcRhsuAI), and place them in `yolov3/weights/`
> 2. **Update `.cfg`** file. Each YOLO layer outputs 255 values: 85 values [4 box coordinates + 1 object confidence + 80 class confidences] per anchor, times 3 anchors. For 1 class training, for example, use `[5 + 1] * 3 = 18` outputs. This modification should be made to the output filter preceding each of the 3 YOLO layers. Also modify `classes=80` to `classes=1` in each YOLO layer. Save as `yolov3-1cls.cfg`
> 
> <img alt="screenshot 2019-02-21 at 19 40 01" width="723" src="https://user-images.githubusercontent.com/26833433/53193125-84637d00-3610-11e9-8b21-e3da8ae2b721.png">
> 
> 1. **Train.** Run `python3 train.py --cfg cfg/yolov3-1cls.cfg --transfer`. You should observe in the Model Summary that only the 3 YOLO layers have their gradients activated now:
> 
> <img alt="Screenshot 2019-04-03 at 11 26 24" width="801" src="https://user-images.githubusercontent.com/26833433/55468450-abe03900-5603-11e9-9c45-fd83183c38b8.png">

**Comment by ganLover** [comment]
> Just use the repo you cited with `python3 train.py --transfer`. How much easier can it get?
> https://github.com/ultralytics/yolov3
> 
> > This guide explains how to train your data with YOLOv3 **using Transfer Learning**. Transfer learning can be a useful way to quickly retrain YOLOv3 on new data without needing to retrain the entire network. We accomplish this by starting from the official YOLOv3 weights, and setting each layer's `.requires_grad` field to false that we do not want to calculate gradients for and optimize.
> > ## Before You Start
> > 
> > 1. Update software (Python >= 3.7, Pytorch >= 1.0, etc.)
> > 2. Clone latest repo: `git clone https://github.com/ultralytics/yolov3`
> > 3. Download COCO dataset: `bash yolov3/data/get_coco_dataset.sh`
> > 
> > ## Transfer Learning
> > 
> > 1. **Download the official YOLOv3 weights** `yolov3.pt` (converted to PyTorch) from our [Google Drive folder](https://drive.google.com/drive/folders/1uxgUBemJVw9wZsdpboYbzUN4bcRhsuAI), and place them in `yolov3/weights/`
> > 2. **Update `.cfg`** file. Each YOLO layer outputs 255 values: 85 values [4 box coordinates + 1 object confidence + 80 class confidences] per anchor, times 3 anchors. For 1 class training, for example, use `[5 + 1] * 3 = 18` outputs. This modification should be made to the output filter preceding each of the 3 YOLO layers. Also modify `classes=80` to `classes=1` in each YOLO layer. Save as `yolov3-1cls.cfg`
> > 
> > <img alt="screenshot 2019-02-21 at 19 40 01" width="723" src="https://user-images.githubusercontent.com/26833433/53193125-84637d00-3610-11e9-8b21-e3da8ae2b721.png">
> > 
> > 1. **Train.** Run `python3 train.py --cfg cfg/yolov3-1cls.cfg --transfer`. You should observe in the Model Summary that only the 3 YOLO layers have their gradients activated now:
> > 
> > <img alt="Screenshot 2019-04-03 at 11 26 24" width="801" src="https://user-images.githubusercontent.com/26833433/55468450-abe03900-5603-11e9-9c45-fd83183c38b8.png">

Thanks for Support

**Comment by selimarikan** [comment]
No need to use another repo. Here is how I am doing the transfer learning:

I have this function that prepares the model. It freezes all the layers first up to the provided index.
Then I just create new layers to replace existing ones with correct parameters (class count, attribute count, etc.).
This is a sample for YOLOv3 Tiny but you can easily adapt it for normal YOLO as well.

Another small thing is to visualize the network to find out which layers you want to freeze. I am using [netron](https://github.com/lutzroeder/netron) tool for that. Just load the .CFG file and you will see the network model.

```
def prepare_pretrained(model, class_count, freeze_index):
    # Modify the network, first freeze all layers
    for name, param in model.named_parameters():
        # module_list.21.batch_norm_21.bias
        layer_id = int(name.split('.')[-3])

        # Freeze layers before 6
        if layer_id < freeze_index:
            param.requires_grad = False

    # For YOLOv3 Tiny, edit these:
    # 23-> yolo_23 -> Last YOLO -> Set classes
    # 22-> conv_22 -> Last Conv -> Set output
    # 16-> yolo_16 -> Prev YOLO ->
    # 15-> conv_15 -> Prev Conv

    # Bounding box attribute count calculation
    # This value is used both in YOLO and last conv layers
    # [4 box coordinates + 1 object confidence + 80 class conf] -> 255 = 85 * 3
    # [4 box coordinates + 1 object confidence + 1 class conf]  -> 18  =  6 * 3
    bb_attrib_count = (5 + class_count) * 3

    # 1- Set last YOLO layer output size
    yolo_23 = model.module_list[23][0]
    new_yolo_23 = YOLOLayer(yolo_23.anchors, class_count, yolo_23.image_dim)
    model.module_list[23][0] = new_yolo_23

    # 2- Set the out filters of conv before last YOLO
    conv_22 = model.module_list[22][0]
    new_conv_22 = torch.nn.Conv2d(
            in_channels=conv_22.in_channels,
                                  out_channels=bb_attrib_count,
                                  kernel_size=conv_22.kernel_size,
                                  stride=conv_22.stride,
                                  padding=conv_22.padding)
    model.module_list[22][0] = new_conv_22
    ...
    ...
```

**Comment by Serhiy-Shekhovtsov** [comment]
Thanks for the example @selimarikan, it's very helpful. I just wanted to add my two cents here, `image_dim` property is now called `img_dim`. So, you can create new YOLO layer like this:
`YOLOLayer(yolo_23.anchors, class_count, yolo_23.img_dim)`
Also, probably you will want to move new layers' parameters to a target device:
```
model.module_list[23][0] = new_yolo_23.to(device)
...
model.module_list[22][0] = new_conv_22.to(device)
```

**Comment by Flova** [comment]
I closed this issue due to inactivity. Feel free to reopen for further discussion.
